### PR TITLE
Allow artifact_file to handle snapshot and latest version correctly (V2 !)

### PR DESCRIPTION
**Pull request by BarthV**
Hi !

I've cleaned up my previous PR and I have taken into account the comments and requests.
I successfully tested many update/upgrade/downgrade/init scenarios from Nexus.

Please test and review this work.

PS : I know some `if` tests may be quite long ... Perhaps it would be interesting to simplify this into a `def`

### CODE REVIEW

**Comment by BarthV** [comment]
Target file is now renamed with provided name during the process.

File: libraries/chef_artifact_nexus.rb [hunk 1]
```diff
@@ -55,6 +55,34 @@ def retrieve_from_nexus(source, destination_dir)
...
+        ::File.basename(REXML::Document.new(remote.get_artifact_info(coordinates)).elements["//repositoryPath"].text)
+      end
+
+      # Returns true when the coordinates specifies a snapshot version.
+      #
+      # @param  location [String] the artifact_location
+      #
+      # @return [Boolean] true when the coordinates's version is a snapshot.
+      def snapshot?(coordinates)
+        coordinates.split(':')[-1].include?('-SNAPSHOT')
```
Comment (KAllan357) [inline libraries/chef_artifact_nexus.rb]: I still think the interim step here is important. Assuming that the error handling is good, we should avoid this array splicing and just use 

```
artifact = NexusCli::Artifact.new(coordinates)
aritfact.version.include?('-SNAPSHOT')
```

File: libraries/chef_artifact_nexus.rb [hunk 2]
```diff
@@ -55,6 +55,34 @@ def retrieve_from_nexus(source, destination_dir)
...
+      end
+
+      # Returns true when the coordinates specifies latest version.
+      #
+      # @param  location [String] the artifact_location
+      #
+      # @return [Boolean] true when the coordinates's version is latest.
+      def latest?(coordinates)
+        coordinates.split(':')[-1].casecmp('LATEST') == 0
```
Comment (KAllan357) [inline libraries/chef_artifact_nexus.rb]: Similar comment as above.

File: providers/file.rb [hunk 3]
```diff
@@ -58,9 +58,18 @@ def load_current_resource
         run_proc :after_download
       end
     elsif Chef::Artifact.from_nexus?(file_location)
-      unless ::File.exists?(new_resource.path) && checksum_valid?
+      unless ::File.exists?(new_resource.path) && checksum_valid? && (!nexus_connection.snapshot?(file_location) || !nexus_connection.latest?(file_location))
         begin
-          nexus_connection.retrieve_from_nexus(file_location, ::File.dirname(new_resource.path))
+          if ::File.exists?(new_resource.path)
+            if Digest::SHA256.file(new_resource.path).hexdigest != nexus_connection.get_artifact_sha(file_location)
```
Comment (KAllan357) [inline providers/file.rb]: I'm not super sure this is working correctly. Nexus only does SHA1 checksums and this is a comparison between a 256 and a 1

**Comment by BarthV** [comment]
Roger that, I'll update this 2 tests using artifact method calls

**Comment by BarthV** [comment]
I'll verify checksuming tomorrow, retest and fix if needed. (It was OK when i tested this so far .. But I could have missed something)
Thx for the review btw

**Comment by BarthV** [comment]
updated.
Previous comments are now fixed.

File: providers/file.rb (hunk 3 above)
Comment (BarthV) [inline providers/file.rb]: indeed ... I made a mistake here.
Thanks for pointing this.

File: libraries/chef_artifact_nexus.rb (hunk 2 above)
Comment (BarthV) [inline libraries/chef_artifact_nexus.rb]: Deleted in favor of artifact class embedded methods

**Comment by KAllan357** [comment]
This looks good for now, though I do agree that the `if` statements are getting a bit crazy. :+1:, but I think I'd like to see us revisit this whole Provider and see if it can be streamlined a bit.

TYVM for the contribution!

**Comment by bdemers** [comment]
Works Here! Got an ETA on the next release ?

**Comment by KAllan357** [comment]
I'd like to see most of the 2.0.0 milestones closed before we cut a release. To me, the big ones are #113, #57 (in general, some updates / support for Chef 11). I also thought I had an issue to refactor the data bag loading logic to simplify it, but I can't seem to find that.

**Comment by BarthV** [comment]
Do you need some help to work on #113 & #57 ?
I'll be happy to have a new release for this CB too :smiley_cat:

**Pull request merged**